proposal/nested-map-functions.md

```diff
@@ -0,0 +1,322 @@
+Projects often build thir own tooling.
+
+The results are inconsistent across projects, difficult to re-use, and often
+slow to compile. Implementing core support for nested maps could change all that.
+
+## Summary
+
+> This section is non-normative.
+
+This proposal updates existing map functions with better support for inspection
+and manipulation of nested maps, as well as adding new functions to the
+`sass:map` module. For existing legacy functions (`get()`, `has-key()`,
+`merge()`) the new behavior will be accessible through both the `sass:map`
+module, and global legacy names (`map-get()`, `map-has-key()`, `map-merge()`).
+New functions (`set()`, `deep-merge()`) will only be available inside the
+`sass:map` module.
+
+The `has-key()` and `get()` functions both accept multiple `$keys...`:
+
+```scss
+@use 'sass:map';
+
+$nav: (
+  'bg': 'gray',
+  'color': (
+    'hover': (
+      'search': yellow,
+      'home': red,
+      'filter': blue,
+    ),
+  ),
+);
+
+$has-search: map.has-key($nav, 'color', 'hover', 'search'); // true
+$search-hover: map.get($nav, 'color', 'hover', 'search'); // yellow
+```
+
+The `merge()` function now accepts multiple `$keys...` between the two maps
+being merged. The keys form a path to the nested location in `$map1` where
+`$map2` should be merged. For example, we update the hover colors in our `$nav`
+map above:
+
+```scss
+@use 'sass:map';
+
+$new-hover: (
+  'search': green,
+  'logo': orange,
+);
+
+$nav: map.merge($nav, 'color', 'hover', $new-hover);
+
+// $nav: (
+//   'bg': 'gray',
+//   'color': (
+//     'hover': (
+//       'search': green,
+//       'home': red,
+//       'filter': blue,
+//       'logo': orange,
+//     ),
+//   ),
+// );
+```
+
+This proposal also adds a `set()` function to `sass:map`, with a similar syntax,
+returning a map with any nested key set to a specific value. To achieve the
+same output as our merge example, we can set each key individually:
+
+```scss
+@use 'sass:map';
+
+$nav: map.set($nav, 'color', 'hover', 'search', green);
+$nav: map.set($nav, 'color', 'hover', 'logo', orange);
+```
+
+And finally, a new `deep-merge()` function in the `sass:map` module allows
+merging two or more nested maps. This works much like the existing `merge()`
+function, but when both maps have a nested-map at the same key, those nested
+maps are also merged:
+
+```scss
+@use 'sass:map';
+
+$nav: (
+  'bg': 'gray',
+  'color': (
+    'hover': (
+      'search': yellow,
+      'home': red,
+      'filter': blue,
+    ),
+  ),
+);
+
+$update: (
+  'bg': white,
+  'color': (
+    'hover': (
+      'search': green,
+      'logo': orange,
+    )
+  )
+);
+
+$nav: map.deep-merge($nav, $update);
+
+// $nav: (
+//   'bg': white,
+//   'color': (
+//     'hover': (
+//       'search': green,
+//       'home': red,
+//       'filter': blue,
+//       'logo': orange,
+//     ),
+//   ),
+// );
+```
+
+## Functions
+
+### `get()`
+
+```
+get($map, $keys...)
```

Comment:
- Is it strange that this new syntax (on several functions) uses an arglist of keys to represent _nested_ keys, while the existing `map.remove()` uses the same syntax to represent _multiple_ keys? 
- Is it strange that we have no nested-key `remove()` syntax?

Reply:
Yes and yes. I can see a few ways to improve this moving forward:

* Deprecate the existing `map.remove()` behavior, and after a suitable period replace it with new behavior. This represents a pretty painful breaking change, though—we usually don't change behavior in a way that makes old code do something *different* than it used to unless we absolutely have to.

* We could add a new function, `map.delete()`, which behaves consistently with the rest of the new map functions. After a period of time we'd deprecate `map.remove()`, and then eventually we'd remove it entirely (or rename it `map.remove-all()`?). The major downside of this strategy is that it's completely non-obvious to a casual observer what the difference between `map.delete()` and `map.remove()` is—they would *have* to read the documentation to understand.

* We could add a new function, `map.deep-remove()`, which behaves consistently with the rest of the new map functions, but plan to leave `map.remove()` as-is.

* We could do nothing at all, and have `map.remove()` just be an odd function out.

I'm not sure which of these options is best.